### DISCUSSION TITLE
License of generated boilerplate

### DISCUSSION THREAD

**Discussion by matheburg** [pyscaffold/pyscaffold#459] [Q&A] [answered]
Given I have generated boilerplate documentation, configuration and code with `pyscaffold`. What are the legal/license conditions for these artifacts?

To give an example, I have a generated `.coveragerc` file that comes along with useful comments that help to understand the configuration. I want to keep them, but I fear I would have to comply with obligations that I am not even aware of.

**Comment by abravalheri** [comment]
Hi @matheburg, I am not a lawyer myself, so I cannot guarantee things for sure, but PyScaffold is licensed itself under the MIT license, which is a very permissive license.

The code generated by PyScaffold is contained in its own source code (under its  [templates](https://github.com/pyscaffold/pyscaffold/tree/master/src/pyscaffold/templates) folder), and therefore also subject to the same MIT license.

This should mean (and is the intention of the project) that you keep the files generated and assign to them whichever other license you want 😄 (including proprietary). I think the MIT license is  very compatible with that. As a safeguard for the PyScaffold devs, the only thing that the MIT license implies is:

```
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR 
PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE 
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
DEALINGS IN THE SOFTWARE.
```

I.e. when you use PyScaffold in your projects, the PyScaffold team is not responsible/liable for it.

Please also be aware that pieces of code generated by third party extensions might differ in terms of licensing. But in general, project scaffolding solutions have a compatible license.

**Reply by matheburg** [reply]
Hi @abravalheri,

Thanks for the fast feedback. This sounds reasonable to me. One should just be aware that this also means all projects that use `pyscaffold` for boilerplate once would have to attribute `pyscaffold`. I don't think this is really transparent. And it may even hold back companies from using `pyscaffold` as they do not want to have scattered FOSS in their source code.

Do you think one could change the templates to public domain (see also https://twitter.com/FlorianWilhelm/status/1405083470574018562)?

**Reply by abravalheri** [reply]
Hi @matheburg, thanks for the feedback.

I wonder if the MIT license implies that you have to attribute the generated code to PyScaffold... For sure if someone copy internal parts of the code that make PyScaffold itself, that would be the case, but for the generated files, it seems that it is not the general understanding.

It seems that other projects use a similar approach: the templates are contained inside the library which have a permissive license. For example, [django-admin](https://github.com/django/django/tree/main/django/conf) and [cookiecutter](https://github.com/audreyfeldroy/cookiecutter-pypackage) have BSD-3 licenses (which is very similar to MIT) and contain inside their repositories the templates used for generating code. Both don't seem to make any special mention in the license.

The same is true for communities other than Python that rely heavily in code generators, such as [Rails](https://github.com/rails/rails), [create-react-app](https://github.com/facebook/create-react-app), [ember-cli](https://github.com/ember-cli/ember-cli); or even official project/package generators from different programming languages, such as [mix for Elixir](https://github.com/elixir-lang/elixir/blob/master/lib/mix/lib/mix/tasks/new.ex) and
[bundler for Ruby](https://github.com/rubygems/rubygems/tree/master/bundler/lib/bundler/templates)

**Reply by FlorianWilhelm** [reply]
I am also not 100% sure about this but if it raises concerns we could make it explicit. My suggestion would be to explicitly state that all files in `pyscaffold.templates` are Public Domain and so are the generated files. 
One could do this by:
* adding a note within `pyscaffold.templates.__init__.py`,
* and adding a gist of this conversation as a new entry to the FAQ.

What do you think @abravalheri?

**Reply by abravalheri** [reply]
Sounds good to me.

I would suggest to be more specific than "Public Domain" and use [BSD-0](https://opensource.org/licenses/0BSD) for the generated code.
I think the folks behind the BSD organisations have explored a lot in the last decades the implications, and it should be legally sound. It seems to be acceptable even for the [Google lawyers](https://opensource.google/docs/patching/#forbidden)

**Reply by abravalheri** [reply]
We could also mention in a preamble in the LICENSE file that all the generated code is under 0-BSD.

In a related note: [Django](https://github.com/django/django/blob/main/LICENSE.python) uses 2 files for its licenses

**Reply by FlorianWilhelm** [reply]
Cool, then we have a solution. Thanks, @abravalheri for finding the example from Django.

* Preamble in the LICENSE file that all the generated code is under 0-BSD.
* Note within `pyscaffold.templates.__init__.py` that templates are 0-BSD.
* gist of this conversation as a new entry to the FAQ.

**Reply by FlorianWilhelm** [reply]
@abravalheri I'll make a proposal PR.

**Reply by abravalheri** [reply]
Thank you very much @FlorianWilhelm

**Comment by FlorianWilhelm** [comment, marked as answer]
@matheburg, it's changed in PR #461 and the RTD docs (latest) show the changes already. With the next version, it will also land in stable.

**Comment by FlorianWilhelm** [comment]
@abravalheri it seems like the license parser of Github is now confused about the preamble of the `LICENSE.txt` file. 
![image](https://user-images.githubusercontent.com/3184072/122668604-2b448d80-d1b9-11eb-94a3-26edee42b8e4.png)
Now it states: "View license" and before it was "MIT".


I don't know exactly if this is a problem. Do we use any services that require PyScaffold to not only be public but also under an OSS license? If these services just use the Github API we might run into a problem.

**Reply by abravalheri** [reply]
Umm, I see. I am not aware about these kind of services. I thibk everything we use only require the GitHub repo to be public.

Bu in the PR for the copyright notice yesterday, I included a few examples from the Apache foundation. I noticed in their projects they have a LICENSE file and a NOTICE file and in the notice file they include parts of the projects covered by other license and license transfers

We could move the preamble there and separate the license files as Django do...

**Reply by abravalheri** [reply]
By the way, I also had a look around and the GitHub license recognizer is very sensitive... Django for example have a BSD3 license that is also reported as "Other".

I don't think tool writers can afford to rely on it...
